[FIX #3706] collect mixpanel events while offline

msuess opened the pull request:
fixes #3706 

### Summary:

Collect mixpanel events on request failure and retry on new event using batch endpoint.

Current implementation:
- if online, submit event to mixpanel
- if offline, enqueue event
- when the next event happens, try to submit the new event and the accumulated ones to mixpanel (in batches of 50, as that is the maximum number of events the mixpanel endpoint accepts)

Edge cases:
- Since we wait for a new event before trying to submit again, we might lose events if the user closes the app or the app gets killed

Possible improvements:
- Instead of waiting until the next event occurs, use a timer to periodically flush the event queue to mixpanel. However, this adds complexity because it requires a go-loop that is constantly running (the current solution only spawns a temporary go-loop for draining the queue)

It would be great if some of the core contributors could chime in with some guidance.

Thanks, I hope this makes sense!

### Steps to test:
- open Status
- turn on airplane mode
- tap around in the app
- turn off airplane mode
- tap once more
- see the events in mixpanel

<!-- (PRs will only be accepted if squashed into single commit.) -->

status: ready

msuess committed:
[ISSUE #3706] collect mixpanel events while offline

This adds a tracking interceptor to `default-interceptors` that replaces the old
`add-post-event-callback` hook. The interceptor gets the required data from
app-db to know if it's online and passes that to a new mixpanel tracking fn that
queues up tracking events, and drains that queue iff the app is online again.

Signed-off-by: Julien Eluard <julien.eluard@gmail.com>

jeluard commented on src/status_im/utils/mixpanel.cljs:
```diff
@@ -1,35 +1,56 @@
+(async/go-loop [accumulator []]
+  (let [accumulated (conj accumulator (<! pending-events-queue))
+        next-accumulator (async/chan)]
+    (-> (rn-dependencies/fetch
```
Any reason we can stick with our `http` abstraction?

msuess replied:
AFAICT the `post` function in the http abstraction has the status server host hardcoded into it, see [here](https://github.com/status-im/status-react/blob/d305f5a4a0afee92e8e72935b972380e3ed33adc/src/status_im/utils/http.cljs#L17).

jeluard replied:
Good catch! We certainly don't want that ..

jeluard commented on src/status_im/utils/mixpanel.cljs:
```diff
@@ -1,35 +1,56 @@
+
+(def ^:private pending-events-queue (async/chan 2000))
+
+(async/go-loop [accumulator []]
```
Id'd rather have the loop only used when the event can't be sent. The happy path would then be the raw http outside async, if the event can't be sent it's added in a queue.

jeluard replied:
Similarly we want to control the lifecycle of the loop (start/stop). Having an always running loop is problematic in the case of switching users for instance.

msuess replied:
Will do!

dmitryn replied:
what do you guys think about using `mount` or `integrant` to manage stateful components?

dmitryn replied:
https://github.com/tolitius/mount/blob/master/doc/clojurescript.md#managing-state-in-clojurescript
https://github.com/weavejester/integrant

jeluard replied:
It's definitively something we should consider. Would help with startup sequence and probably testing too.

jeluard reviewed (changes requested):
There is a way to know status is offline (see `offline?` sub). It makes sense to queue only in this case, and to fire directly otherwise.
It's ok to miss some events if mixpanel is down for instance.
So I wouldn't hook on failed http calls for instance.

msuess commented:
Thanks for the feedback @jeluard, will be looking into this!

I am aware of the `offline` sub but I wasn't sure how to use it outside of effect handlers. Since the tracking gets fired after every re-frame event via `add-post-event-callback` I'm not sure where I could listen for this sub. Do you have any ideas how to solve this cleanly?

dmitryn commented on src/status_im/utils/mixpanel.cljs:
```diff
@@ -1,35 +1,56 @@
+    (-> (rn-dependencies/fetch
+         base-track-url
+         (clj->js {:method  "POST"
+                   :body    (str "data=" (encode (into [] (take 50 accumulated))))
```
i would move `50` to something like `def batch-size` (same with `50` on line 44)

msuess replied:
Good catch, will do!

jeluard commented:
@msuess Great point I am not sure how to do that. @flexsurfer @dmitryn Any idea how to address @msuess previous point?

dmitryn commented:
>Any idea how to address @msuess previous point?

I'm not aware of clean way to access re-frame app-db outside of event handler or subscription. Dirty hack would be setting network-status as a global var.

Second way to solve this: introduce re-frame event like `send-tracking-events` which will be fired/dispatched in a loop every N ms and will be actually sending events (in batches) or skipping iteration if app is offline.

jeluard commented:
@msuess Might be worth checking on clojurians/re-frame slack channel. Also it might be a case of discussing changes to `add-post-event-callback` so that we get a handle on `app-db` or something along those lines.

msuess commented:
I posted my question in the slack. Let's see what happens.

msuess commented:
According to the answer I got on slack, we could try adding an [interceptor](https://github.com/Day8/re-frame/blob/master/docs/Interceptors.md) for this to every one of our event handlers. Luckily, we're already using custom wrappers for `reg-event-db` and `reg-event-fx` and could add the relevant code in [utils/handlers.cljs](https://github.com/status-im/status-react/blob/38c9bb7266a05778a61e3e7fbb41d121a7831209/src/status_im/utils/handlers.cljs#L90). What do you think @jeluard?

jeluard commented:
@msuess Might make sense to use interceptors. I wonder what are the tradeoffs compared to using `add-post-event-callback`.

msuess commented:
@jeluard I went ahead and implemented a solution using an interceptor, please let me know what you think.

@dmitryn feel free to check out the new implementation if you're interested :)

jeluard commented on src/status_im/utils/handlers.cljs:
```diff
@@ -81,14 +83,37 @@
+       (when (get-in new-db [:accounts/accounts
+                             current-account-id
+                             :sharing-usage-data?])
+         (let [event (get-coeffect context :event)
```
Can you align values in map?

msuess replied:
You mean in `let`? Can do!

jeluard commented on src/status_im/utils/mixpanel.cljs:
```diff
@@ -1,35 +1,73 @@
+(defn- submit-batch [events]
+  (let [done-chan (async/chan)]
+    (-> (rn-dependencies/fetch
+         base-track-url
```
Can we reuse our post abstraction? It's ok if it has to be modified.

msuess replied:
Sure!

jeluard commented on src/status_im/utils/mixpanel.cljs:
```diff
@@ -1,35 +1,73 @@
+(def max-batch-size 50)
+
+(defn- drain-events-queue! []
+  (let [done-chan (async/chan)]
```
Some comment would be appreciated :) Also can this be tested?

msuess replied:
Will add some comments and tests.

jeluard commented on src/status_im/utils/mixpanel.cljs:
```diff
@@ -1,35 +1,73 @@
+    (if offline?
+      (async/go (async/>! pending-events-queue event))
+      (async/go
+        (async/<! (drain-events-queue!))
```
So we try to drain on next online event? What if we get back online but no new event are generated?

msuess replied:
Then the channel will be immediately closed and nothing happens, as `async/alts!` will return `::drained`.

msuess replied:
To clarify: we try and drain on the next tracking event that happens while online.

jeluard replied:
I guess that's good enough. Can you add a comment here too? :)

msuess replied:
Sure!

cammellos replied:
Hi @msuess, thanks for the PR, looking really good!

I was wondering maybe there are a few things we can simplify:

We probably don't need `done-chan`, as `go-loop` will return a chan which will be populated with the return value of the loop, and closed after:

`(a/<!! (go-loop [] (Thread/sleep 2000))` waits for 2 seconds

You can also use `poll!` instead of `alts!` and a default value (poll! will return nil if the channel is empty)

We also tend to use `seq` to check if a collection is not empty

something like:

```
(defn- drain-events-queue! []
  (async/go-loop [accumulator []]
    (let [event (async/poll! pending-events-queue)]
      (cond
        (nil? event)
        (when (seq accumulator)
              (async/<! (submit-batch accumulator)))

        (= (- max-batch-size 1) (count accumulator))
        (do (async/<! (submit-batch (conj accumulator event)))
            (recur []))
        :else
        (recur (conj accumulator event))))))
```

I would personally also use `partition-all`, which makes the code a bit more compact (there some cpu tradeoff, but don't think has an impact):
```
(defn- drain-events-queue! []
  (let [events (loop [accumulator []] ; or (take-while (fn [_] (a/poll! c)) (range)) which returns a lazy-seq
                 (if-let [event (async/poll! pending-events-queue)]
                   (recur (conj accumulator event))
                   accumulator))]
    (async/go
              (doseq [batch (partition-all max-batch-size events)] 
                (async/<! (submit-batch batch))))))
```

`partion-all` can also be applied as a transducer to the channel `(chan 100 (partition-all 2))`, which means we would not have to loop and recur as elements would be batched, but unfortunately we would need to close the channel to process everything, I can't think of a better way, but if you can think of a neat solution go for it!

Not sure about this one, but we could also always push to the queue, and drain when we are online:

```
  (async/go (async/>! pending-events-queue event))
  (when-not offline?
   (async/go
    (async/<! (drain-events-queue!)) ; without the http/get

```

what do you think?

jeluard reviewed (commented):
Nice work! i left some comments.

msuess commented:
Wow, @cammellos, that's amazing input! Thank you so much!

msuess commented:
@jeluard  @cammellos  I updated the PR according to your input.

I came up with a test for `drain-events-queue!`, but it relies on atoms and is kind of ugly. This is the best I could come up with without making the `drain-events-queue!` implementation more clunky, but maybe you guys have a better idea as to how to test this?

I'm always pushing to the queue now as per @cammellos' suggestion because I think it makes the implementation simpler. @jeluard let me know if you're fine with this or if I should use a normal `get` in case the app is online.

I changed the `http/post` abstraction as well and removed the `:http-post` fx handler and respective constants for the status api server address since it's not being used.

I've added some docstrings and comments, let me know if you think this is sufficient or too much @jeluard.

jeluard commented on test/cljs/status_im/test/utils/mixpanel.cljs:
```diff
@@ -1,20 +1,41 @@
-  (is (= 1 (count (mixpanel/matching-events [:key] definitions))))
-  (is (= 2 (count (mixpanel/matching-events [:key :subkey] definitions))))
-  (is (empty? (mixpanel/matching-events [:key1 :another-subkey] definitions))))
+(t/deftest matching-event
```
Any reason to introduce `t/`

msuess replied:
I introduced it because I needed to import "async" and I remembered that the style guide says that I should [avoid `:refer`](https://wiki.status.im/Clojure_Style_Guide#Use_of_:as_and_:refer). I can change it back if you want.

cammellos replied:
@msuess that's actually a good point! In tests we do use `:refer` as it's more idiomatic, we need to update the style guide. Good spot, thanks.

msuess replied:
Alright, I'll change it back then :)

cammellos replied:
thanks, sorry for the confusion

jeluard commented:
@msuess Looks good to me! The only question I have left is related to your ns change in tests. Any reason to do that?

cammellos commented on test/cljs/status_im/test/utils/mixpanel.cljs:
```diff
@@ -1,20 +1,41 @@
+   (let [queue (async/chan (async/sliding-buffer 2000))
+         results (atom [])]
+     (async/go
+       (async/<! (async/onto-chan queue (into [] (take 123 (range))) false))
```
this can be

```
(async/onto-chan queue (range 123) false)
```

cammellos commented on test/cljs/status_im/test/utils/mixpanel.cljs:
```diff
@@ -1,20 +1,41 @@
+         queue
+         (fn [events]
+           (let [result-chan (async/chan)]
+             (swap! results conj (into [] events))
```
Same here:
``` 
(swap! results conj events)
```

cammellos commented on test/cljs/status_im/test/utils/mixpanel.cljs:
```diff
@@ -1,20 +1,41 @@
+             (swap! results conj (into [] events))
+             (async/go (async/close! result-chan))
+             result-chan))))
+       (t/is (= @results [(into [] (range 50))
```
and here:
`[(range 50)...]`

msuess commented:
Thanks @cammellos, will fix!

msuess commented:
@jeluard @cammellos I implemented the changes you requested :)

jeluard commented:
Awesome thank you! Sorry it took so many back and forth :(

lukaszfryc commented:
Works great 👍 I tested different scenarios related to offline mode, logging out, using different accounts, checking if events are not duplicated, etc.

dmitryn commented on src/status_im/utils/mixpanel.cljs:
```diff
@@ -1,38 +1,81 @@
+                 (log/debug "successfully submitted events")
+                 (async/go (async/close! done-chan)))
+               (fn [error]
+                 (log/error "error while submitting events" error)
```
Are you sure `log/error` doesn't bring react-native "redbox" error to the user? Not sure user wants to know about such error.

msuess commented:
Thanks everyone!
